compressors/fano_coder.py

```diff
@@ -0,0 +1,154 @@
+"""Fano Coding
+Shannon coding is a basic prefix free coding. There is some confusion in literature about Shannon, Fano and
+Shannon-Fano codes, e.g. see Naming section of https://en.wikipedia.org/wiki/Shannon–Fano_coding.
+
+This document uses Fano coding as described in the wiki article above.
+"""
+from typing import Any, Tuple
+from utils.bitarray_utils import BitArray
+from utils.test_utils import get_random_data_block, try_lossless_compression
+from compressors.prefix_free_compressors import PrefixFreeTree, PrefixFreeEncoder, PrefixFreeDecoder, BinaryNode
+from core.prob_dist import ProbabilityDist
+
+
+class FanoTree(PrefixFreeTree):
+    def __init__(self, prob_dist):
+        super().__init__(prob_dist)
+        # sort the probability distribution
+        self.sorted_prob_dist = ProbabilityDist.get_sorted_prob_dist(prob_dist.prob_dict)
+        # initialize root node of Fano Tree
+        self.root_node = BinaryNode(id=None)
+
+        # build tree returns the root node
+        super().__init__(root_node=self.build_fano_tree(self.root_node, self.sorted_prob_dist))
+
+    @staticmethod
+    def normalize_prob_dict(prob_dict):
+        """
+        normalizes subset of original probability dict to 1
+        """
+        sum_p = sum(prob_dict.values())
+        return ProbabilityDist(dict([[a, b / sum_p] for a, b in prob_dict.items()]))
+
+    @staticmethod
+    def criterion(dict_item):
+        """used for finding the symbol at which cumulative probability is nearest to 0.5
+        This symbol is used to split between left and right trees of Fano. We split right starting this symbol
+        to get most balanced tree.
+        """
+        key, value = dict_item
+        return abs(value - 0.5)
+
+    @staticmethod
+    def build_fano_tree(root_node, norm_sort_prob_list) -> BinaryNode:
+        """recursively build Fano Tree"""
+        # Get cumulative probability dict
+        cumulative_prob_dict = norm_sort_prob_list.cumulative_prob_dict
+
+        # Split the tree into left and right nodes based on cumulative probability and recursively build
```

Comment:
Once you move some logic to `split_prob_dist_into_two`, this function can be simplified as:

```python
    def build_fano_tree(root_node, norm_sort_prob_dist):
        # base case: 
        if only one node: 
           root_node.id = ...
           return root_node
        else:
           # split norm_sort_prob_dist into two normalized prob dists
           left_prob_dist, right_prob_dist = split_prod_dist_into_two(norm_sort_prob_dist)
           
            # call recursion on left node
           root_node.left_node = BinaryNode(id="None", ...)
           build_fano_tree(root_node.left_node, left_prob_dist)
           
           # call recursion on right node
           root_node.left_node = BinaryNode(id="None", ...)
           build_fano_tree(root_node.left_node, left_prob_dist)
           
           return root_node
        
```

Reply:
I have not checked whether this works, but you can try to get it working..
In general, it is better to start with writing a pseudo-code like structure and then filling in the details

Reply:
this has some issue. happy to discuss during our call today.